sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultStringTypes.scala

```diff
@@ -0,0 +1,201 @@
+   * new type instead of the default string type.
+   */
+  private def transformPlan(plan: LogicalPlan, newType: StringType): LogicalPlan = {
+    plan resolveOperators { operator =>
+      operator resolveExpressionsUp { expression =>
+        transformExpression.applyOrElse((expression, newType), fallbackToExpression)
+      }
+    }
+  }
+
+  private def fallbackToExpression(tuple: (Expression, StringType)): Expression = tuple._1
+
+  /**
+   * Transforms the given expression, by changing all default string types to the given new type.
+   */
+  private def transformExpression: PartialFunction[(Expression, StringType), Expression] = {
+    case (columnDef: ColumnDefinition, newType) if hasDefaultStringType(columnDef.dataType) =>
+      columnDef.copy(dataType = replaceDefaultStringType(columnDef.dataType, newType))
+
+    case (cast: Cast, newType) if hasDefaultStringType(cast.dataType) =>
+      cast.copy(dataType = replaceDefaultStringType(cast.dataType, newType))
+
+    case (Literal(value, dt), newType) if hasDefaultStringType(dt) =>
+      Literal(value, replaceDefaultStringType(dt, newType))
+  }
+
+  private def hasDefaultStringType(dataType: DataType): Boolean =
+    dataType.existsRecursively(isDefaultStringType)
+
+  private def isDefaultStringType(dataType: DataType): Boolean = {
+    dataType match {
+      case _: TemporaryStringType =>
+        !replaceWithTempType
+      case st: StringType =>
+        // should only return true for StringType object and not StringType("UTF8_BINARY")
+        st.eq(StringType)
+      case _ => false
+    }
+  }
+
+  private def replaceDefaultStringType(dataType: DataType, newType: StringType): DataType = {
+    dataType.transformRecursively {
+      case currentType: StringType if isDefaultStringType(currentType) =>
+        if (replaceWithTempType && currentType == newType) {
+          getTemporaryStringType(currentType)
+        } else {
+          newType
+        }
+    }
+  }
+
+  private def getTemporaryStringType(forType: StringType): StringType = {
+    if (forType.collationId == 0) {
+      TemporaryStringType(1)
+    } else {
+      TemporaryStringType(0)
+    }
+  }
+
+  private def replaceColumnTypes(
+      colTypes: Seq[QualifiedColType],
+      newType: StringType): Seq[QualifiedColType] = {
+    colTypes.map {
+      case colWithDefault if hasDefaultStringType(colWithDefault.dataType) =>
+        val replaced = replaceDefaultStringType(colWithDefault.dataType, newType)
+        colWithDefault.copy(dataType = replaced)
+
+      case col => col
+    }
+  }
+}
+
+case object ResolveDefaultStringTypes
+  extends ResolveDefaultStringTypes(replaceWithTempType = true) {}
+
+case object ResolveDefaultStringTypesWithoutTempType
+  extends ResolveDefaultStringTypes(replaceWithTempType = false) {}
+
+case class TemporaryStringType(override val collationId: Int)
+  extends StringType(collationId) {
```

Comment:
this shouldn't extend `StringType`, otherwise `StringType(id) == TemporaryStringType(id)` and we may hit the TreeNode transformation issue again.

Reply:
`getTemporaryStringType` fn makes sure we get a different collation id. I tried not extending StringType but it causes many issues, since for example we have to create a literal with new type and that would of course fail since it doesn't recognize it